bftengine/src/bcstatetransfer/SourceSelector.cpp

```diff
@@ -31,6 +31,16 @@ void SourceSelector::setFetchingTimeStamp(uint64_t currTimeMilli, bool retransmi
 void SourceSelector::removeCurrentReplica() {
   preferredReplicas_.erase(currentReplica_);
   currentReplica_ = NO_REPLICA;
+  receivedValidBlockFromSrc_ = false;
+}
+
+void SourceSelector::onReceivedValidBlockFromSource() {
+  ConcordAssertNE(currentReplica_, NO_REPLICA);
+  if (!receivedValidBlockFromSrc_) {
+    receivedValidBlockFromSrc_ = true;
+    LOG_INFO(logger_, "Insert source " << currentReplica_ << " into actualSources_");
+    actualSources_.push_back(currentReplica_);
```

Comment:
can we give up on the flag ```receivedValidBlockFromSrc_``` if we'll use std::set instead of std::vector for ```actualSources_``` and the presence of currentReplica_ in the set will indicate the same logic as ```receivedValidBlockFromSrc_ == true```.
If we can give up on managing ```receivedValidBlockFromSrc_``` it can save future bugs, if it's not the case, can UT be added ?

Reply:
@cloudnoize , I don't think we can use an STL set here.
The order of the elements is important and a set is associative, and also there can be re-occurring elements.
For example, I can have a vector containing [0,6,1,1,0].

As for the UT suggestion, I think it is a good idea, in practice I wanted to move into test-driven development - so just last week I've inserted the new infra into bcstatetransfer_tests.cpp , Before that it was close to impossible to do your suggestion without a work of at least a week. what was done till now is still very initial. I will see if it is possible to have a unit test with what we have now, else we will have to do it in another task (there is a task for adding unit tests, but it will require weeks of work).
I'll comment again after checking if a unit test can be added..

Reply:
@cloudnoize , I have added a unit-test for the change.
On the way, I've found a minor bug that was fixed as well as part of this PR.